cpp/ql/src/Likely Bugs/UnderspecifiedFunctions/MistypedFunctionArguments.ql

```diff
@@ -0,0 +1,117 @@
+/**
+ * @name Call to a function with one or more incompatible arguments
+ * @description A call to a function with at least one argument whose type does
+ * not match the type of the corresponding function parameter.  This may indicate
+ * that the author is not familiar with the function being called.  Passing mistyped
+ * arguments on a stack may lead to unpredictable function behavior.
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
+ * @id cpp/mistyped-function-arguments
+ * @tags correctness
+ *       maintainability
+ */
+
+import cpp
+
+predicate argTypeMayBePromoted(Type arg, Type parm) {
+  arg = parm
+  or
+  // floating-point and integral promotions
+  arg instanceof FloatType and parm instanceof DoubleType
+  or
+  arg instanceof FloatType and parm instanceof LongDoubleType
+  or
+  arg instanceof DoubleType and parm instanceof LongDoubleType
+  or
+  arg instanceof CharType and parm instanceof IntType
+  or
+  arg instanceof CharType and parm instanceof IntType
+  or
+  arg instanceof CharType and parm instanceof LongType
+  or
+  arg instanceof CharType and parm instanceof LongLongType
+  or
+  arg instanceof ShortType and parm instanceof IntType
+  or
+  arg instanceof ShortType and parm instanceof LongType
+  or
+  arg instanceof ShortType and parm instanceof LongLongType
+  or
+  arg instanceof IntType and parm instanceof LongType
+  or
+  arg instanceof IntType and parm instanceof LongLongType
+  or
+  arg instanceof LongType and parm instanceof LongLongType
+  or
+  arg instanceof Enum and parm instanceof IntType
+  or
+  arg instanceof Enum and parm instanceof LongType
+  or
+  arg instanceof Enum and parm instanceof LongLongType
+  or
+  // enums are usually sized as ints
+  arg instanceof IntType and parm instanceof Enum
+  or
+  // pointer types
+  arg.(PointerType).getBaseType().getUnspecifiedType() = parm
+        .getUnspecifiedType()
+        .(PointerType)
+        .getBaseType()
+        .getUnspecifiedType()
+  or
+  // void * parameters accept arbitrary pointer arguments
+  arg instanceof PointerType and
+  parm.(PointerType).getBaseType().getUnspecifiedType() instanceof VoidType
```

Comment:
I suppose this goes back to (my) understanding what our platform is supposed to do.  In my view, just because some C compiler doesn't warn about passing opaque types to non-opaque functions doesn't mean that we shouldn't.  If we strictly adhered to what C compilers do, the whole <code>MistypedFunctionArguments.ql</code> query would be unnecessary.